bootstrap/kubeadm/internal/ignition/clc/clc.go

```diff
@@ -0,0 +1,377 @@
+		"Split":          strings.Split,
+		"Join":           strings.Join,
+		"MountpointName": mountpointName,
+	}
+}
+
+func mountpointName(name string) string {
+	return strings.TrimPrefix(strings.ReplaceAll(name, "/", "-"), "-")
+}
+
+func templateYAMLIndent(i int, input string) string {
+	split := strings.Split(input, "\n")
+	ident := "\n" + strings.Repeat(" ", i)
+	return strings.Join(split, ident)
+}
+
+func renderCLC(input *cloudinit.BaseUserData, kubeadmConfig string) ([]byte, error) {
+	t := template.Must(template.New("template").Funcs(defaultTemplateFuncMap()).Parse(clcTemplate))
+
+	usersWithPasswordAuth := []string{}
+	for _, user := range input.Users {
+		if user.LockPassword != nil && !*user.LockPassword {
+			usersWithPasswordAuth = append(usersWithPasswordAuth, user.Name)
+		}
+	}
+
+	filesystemDevicesByLabel := map[string]string{}
+	if input.DiskSetup != nil {
+		for _, filesystem := range input.DiskSetup.Filesystems {
+			filesystemDevicesByLabel[filesystem.Label] = filesystem.Device
+		}
+	}
+
+	data := render{
+		BaseUserData:             input,
+		KubeadmConfig:            kubeadmConfig,
+		UsersWithPasswordAuth:    strings.Join(usersWithPasswordAuth, ","),
+		FilesystemDevicesByLabel: filesystemDevicesByLabel,
+	}
+
+	var out bytes.Buffer
+	if err := t.Execute(&out, data); err != nil {
+		return nil, errors.Wrapf(err, "failed to render template")
+	}
+
+	return out.Bytes(), nil
+}
+
+// Render renders the provided user data and CLC snippets into Ignition config.
+func Render(input *cloudinit.BaseUserData, clc *bootstrapv1.ContainerLinuxConfig, kubeadmConfig string) ([]byte, string, error) {
+	if input == nil {
+		return nil, "", errors.New("empty base user data")
+	}
+
+	if clc == nil {
+		return nil, "", errors.New("get empty CLC config")
+	}
+
+	clcBytes, err := renderCLC(input, kubeadmConfig)
+	if err != nil {
+		return nil, "", errors.Wrapf(err, "rendering CLC configuration")
+	}
+
+	userData, warnings, err := buildIgnitionConfig(clcBytes, clc)
+	if err != nil {
+		return nil, "", errors.Wrapf(err, "building Ignition config")
+	}
+
+	return userData, warnings, nil
+}
+
+func buildIgnitionConfig(baseCLC []byte, clc *bootstrapv1.ContainerLinuxConfig) ([]byte, string, error) {
+	// We control baseCLC config, so treat it as strict.
+	ign, _, err := clcToIgnition(baseCLC, true)
+	if err != nil {
+		return nil, "", errors.Wrapf(err, "converting generated CLC to Ignition")
+	}
+
+	var clcWarnings string
+
+	if clc.AdditionalConfig != "" {
+		additionalIgn, warnings, err := clcToIgnition([]byte(clc.AdditionalConfig), clc.Strict)
+		if err != nil {
+			return nil, "", errors.Wrapf(err, "converting additional CLC to Ignition")
+		}
+
+		clcWarnings = warnings
+
+		ign = ignition.Append(ign, additionalIgn)
+	}
+
+	userData, err := json.Marshal(&ign)
+	if err != nil {
+		return nil, "", errors.Wrapf(err, "marshaling generated Ignition config into JSON")
+	}
+
+	return userData, clcWarnings, nil
+}
```

Comment:
Would something like this be fine?
```suggestion
        clcBytes, err := renderCLC(input, kubeadmConfig)
        if err != nil {
                return nil, "", errors.Wrapf(err, "rendering CLC configuration")
        }

        userData, warnings, err := buildIgnitionConfig(clcBytes, clc)
        if err != nil {
                return nil, "", errors.Wrapf(err, "building Ignition config")
        }

        return userData, warnings, nil
}

func buildIgnitionConfig(baseCLC []byte, clc *bootstrapv1.ContainerLinuxConfig) ([]byte, string, error) {
        // We control baseCLC config, so treat it as strict.
        ign, _, err := clcToIgnition(baseCLC, true)
        if err != nil {
                return nil, "", errors.Wrapf(err, "converting generated CLC to Ignition")
        }

        var clcWarnings string

        if clc != nil && clc.AdditionalConfig != "" {
                additionalIgn, warnings, err := clcToIgnition([]byte(clc.AdditionalConfig), clc.Strict)
                if err != nil {
                        return nil, "", errors.Wrapf(err, "converting additional CLC to Ignition")
                }

                clcWarnings = warnings

                ign = ignition.Append(ign, additionalIgn)
        }

        userData, err := json.Marshal(&ign)
        if err != nil {
                return nil, "", errors.Wrapf(err, "marshaling generated Ignition config into JSON")
        }

        return userData, clcWarnings, nil
}
```

Patch version, which is more readable:
```diff
diff --git bootstrap/kubeadm/internal/ignition/clc/clc.go bootstrap/kubeadm/internal/ignition/clc/clc.go
index bbdb51c9b..086251677 100644
--- bootstrap/kubeadm/internal/ignition/clc/clc.go
+++ bootstrap/kubeadm/internal/ignition/clc/clc.go
@@ -314,10 +314,6 @@ func Render(input *cloudinit.BaseUserData, clc *bootstrapv1.ContainerLinuxConfig
                return nil, "", errors.New("empty base user data")
        }

-       if clc == nil {
-               return nil, "", errors.New("get empty CLC config")
-       }
-
        clcBytes, err := renderCLC(input, kubeadmConfig)
        if err != nil {
                return nil, "", errors.Wrapf(err, "rendering CLC configuration")
@@ -340,7 +336,7 @@ func buildIgnitionConfig(baseCLC []byte, clc *bootstrapv1.ContainerLinuxConfig)

        var clcWarnings string

-       if clc.AdditionalConfig != "" {
+       if clc != nil && clc.AdditionalConfig != "" {
                additionalIgn, warnings, err := clcToIgnition([]byte(clc.AdditionalConfig), clc.Strict)
                if err != nil {
                        return nil, "", errors.Wrapf(err, "converting additional CLC to Ignition")
```

Of course tests should follow.

Reply:
make sense to me

Reply:
I'll address it.